### ISSUE TITLE
In the wild inference

### ISSUE THREAD

**Issue by JamesMcCullochDickens** [soeaver/M2FP#4]
Hi, I just read your great survey and I am interested in getting some masks using the mhpv2 model on some images of people not included in any of the human parsing datasets.

Do you have a snippet for loading a pretrained model and doing inference? I will be studying the code to do this in the coming days but any suggestions or pre-existing code would be good. I'll happily share the in the wild inference code after figuring it all out. I've done some stuff with Detectron2 in the past but I will need to dig through some of the repo to really get the hang of it all.

**Comment by JamesMcCullochDickens** [comment]
I have a first attempt here:


```
import torch
from PIL import Image
import numpy as np
import math

import train_net


def get_model(args):
    cfg = train_net.setup(args)
    model = train_net.Trainer.build_model(cfg)
    train_net.DetectionCheckpointer(model, save_dir=cfg.OUTPUT_DIR).resume_or_load(
        cfg.MODEL.WEIGHTS, resume=args.resume)
    model.training = False
    return model


def get_part_color_map(num_vals) -> dict:
    global part_d_num_to_str
    keys = list(range(-1, num_vals))
    color_d = {}
    biases, gain, power = [80, 160, 240], 2, 3
    for key in keys:
        r = int(math.pow((key+biases[0])*gain, power)) % 255.0
        g = int(math.pow((key+biases[1])*gain, power)) % 255.0
        b = int(math.pow((key+biases[2])*gain, power)) % 255.0
        color_d[key] = np.array([r, g, b], dtype=np.uint8)
    return color_d


def mask_to_color_im(mask: np.ndarray) -> np.ndarray:
    color_mask = np.zeros((mask.shape[0], mask.shape[1], 3), dtype=np.uint8)
    color_d = get_part_color_map(20)
    unique_vals = np.unique(mask)
    for val in unique_vals:
        if val == 0:
            continue
        spatial_mask = mask == val
        color_mask[spatial_mask, :] = color_d[val]
    return color_mask


@ torch.no_grad
def inference_on_im(im_fp: str, model: torch.nn.Module):
    pil_im = Image.open(im_fp).resize((800, 800), Image.BILINEAR)
    pil_im_arr = np.asarray(pil_im)
    im_arr = np.asarray(pil_im_arr)/255.0
    im_t = torch.tensor(im_arr).permute((-1, 0, 1)).float()
    d = [{"image": im_t}]
    output = model(d)[0]["parsing"]


    # parts processing -> get one part map
    part_masks = output["part_outputs"]
    part_masks = list(filter(lambda x: x["score"] > 0.10, part_masks))
    part_ids = {i: x["category_id"] for i, x in enumerate(part_masks)}
    all_part_masks = torch.cat([torch.unsqueeze(x["mask"], dim=0) for x in part_masks], dim=0)
    seg_part_mask = torch.argmax(all_part_masks, dim=0)
    for i in range(len(part_masks)):
        seg_part_mask[seg_part_mask == i] = part_ids[i]

    # human mask processing
    human_masks = output["human_outputs"]
    human_masks = list(filter(lambda x: x["score"] > 0.5, human_masks))
    human_masks.sort(key=lambda x: x["score"])
    human_part_masks = []
    for human_mask in human_masks:
        human_mask = torch.sigmoid(human_mask["mask"])
        human_mask[human_mask < 0.5] = 0.0
        human_part_mask = torch.where(human_mask >= 0.5, seg_part_mask, -1)
        human_part_masks.append(human_part_mask.cpu().numpy().astype(int))


    """
    human_instance_mask = torch.sigmoid(human_masks[0]["mask"]).cpu().numpy()
    human_instance_mask[human_instance_mask>=0.65] = 1
    human_instance_mask = human_instance_mask.astype(int)
    human_instance_mask_color = mask_to_color_im(human_instance_mask)
    human_instance_mask_color_im = Image.fromarray(human_instance_mask_color)
    human_instance_mask_color_im.show()
    """

    color_seg_mask = mask_to_color_im(human_part_masks[0])
    overall_im = np.concatenate([pil_im_arr, color_seg_mask], axis=1)
    overall_im = Image.fromarray(overall_im)
    overall_im.show()



if __name__ == "__main__":
    args = train_net.default_argument_parser().parse_args()
    model = get_model(args)
    model = model.to(0)
    sample_im_fp = "/home/jamesdickens/Desktop/Code/HumanParsing/M2FP/sample_person.jpg"
    inference_on_im(sample_im_fp, model)
```

    
    But the results are giving me something like:
    
    
![tmpyflh_8xc](https://github.com/soeaver/M2FP/assets/23439776/e3f84db9-019a-43bc-b61e-af5daf44217d)

Any thoughts about what I might be doing wrong inference wise?

**Comment by JamesMcCullochDickens** [comment]
As a first attempt (note the images are NOT normalized to [0,1], and the channel standardization uses means/stds from raw pixel values). 

```
import torch
from PIL import Image
import numpy as np
import math

import train_net


def get_model(args):
    cfg = train_net.setup(args)
    model = train_net.Trainer.build_model(cfg)
    train_net.DetectionCheckpointer(model, save_dir=cfg.OUTPUT_DIR).resume_or_load(
        cfg.MODEL.WEIGHTS, resume=args.resume)
    model.training = False
    return model


def get_part_color_map(num_vals) -> dict:
    global part_d_num_to_str
    keys = list(range(-1, num_vals))
    color_d = {}
    biases, gain, power = [80, 160, 240], 2, 3
    for key in keys:
        r = int(math.pow((key+biases[0])*gain, power)) % 255.0
        g = int(math.pow((key+biases[1])*gain, power)) % 255.0
        b = int(math.pow((key+biases[2])*gain, power)) % 255.0
        color_d[key] = np.array([r, g, b], dtype=np.uint8)
    return color_d


def mask_to_color_im(mask: np.ndarray) -> np.ndarray:
    color_mask = np.zeros((mask.shape[0], mask.shape[1], 3), dtype=np.uint8)
    color_d = get_part_color_map(20)
    unique_vals = np.unique(mask)
    for val in unique_vals:
        if val == 0:
            continue
        spatial_mask = mask == val
        color_mask[spatial_mask, :] = color_d[val]
    return color_mask


@ torch.no_grad
def inference_on_im(im_fp: str, model: torch.nn.Module):
    pil_im = Image.open(im_fp).resize((800, 800), Image.BILINEAR)
    pil_im_arr = np.asarray(pil_im)
    im_arr = np.asarray(pil_im_arr)
    im_t = torch.tensor(im_arr).permute((-1, 0, 1)).float()
    d = [{"image": im_t}]
    output = model(d)[0]["parsing"]


    # parts processing -> get one part map
    part_masks = output["part_outputs"]
    part_masks = list(filter(lambda x: x["score"] > 0.10, part_masks))
    if not part_masks:
        print(f" No parts detected!")
        part_masks = None
    else:
        part_ids = {i: x["category_id"] for i, x in enumerate(part_masks)}
        all_part_masks = torch.cat([torch.unsqueeze(x["mask"], dim=0) for x in part_masks], dim=0)
        seg_part_mask = torch.argmax(all_part_masks, dim=0)
        for i in range(len(part_masks)):
            seg_part_mask[seg_part_mask == i] = part_ids[i]

    # human mask processing
    human_masks = output["human_outputs"]
    human_masks = list(filter(lambda x: x["score"] > 0.1, human_masks))
    human_masks.sort(key=lambda x: x["score"])
    human_part_masks = []
    if part_masks:
        for human_mask in human_masks:
            human_mask = torch.sigmoid(human_mask["mask"])
            human_mask[human_mask < 0.5] = 0.0
            human_part_mask = torch.where(human_mask >= 0.5, seg_part_mask, -1)
            human_part_masks.append(human_part_mask.cpu().numpy().astype(int))

    if human_masks:
        human_instance_mask = torch.sigmoid(human_masks[0]["mask"]).cpu().numpy()
        human_instance_mask[human_instance_mask>=0.65] = 1
        human_instance_mask = human_instance_mask.astype(int)
        human_instance_mask_color = mask_to_color_im(human_instance_mask)
        human_instance_mask_color_im = Image.fromarray(human_instance_mask_color)
        human_instance_mask_color_im.show()

    # for the moment just showing the first person as a demo, will fix later
    color_seg_mask = mask_to_color_im(human_part_masks[0])
    overall_im = np.concatenate([pil_im_arr, human_instance_mask_color_im, color_seg_mask], axis=1)
    overall_im = Image.fromarray(overall_im)
    overall_im.show()



if __name__ == "__main__":
    args = train_net.default_argument_parser().parse_args()
    model = get_model(args)
    model = model.to(0)
    sample_im_fp = "/home/jamesdickens/Desktop/Code/HumanParsing/M2FP/sample_person2.png"
    inference_on_im(sample_im_fp, model)
```
    
    
![tmp9yhekdcz](https://github.com/soeaver/M2FP/assets/23439776/23509c93-20f6-4b57-9f51-54ab76f4930a)

**Comment by JamesMcCullochDickens** [comment]
Another example just for fun, notice how the model doesn't consider the belt or the arm patch as part of the person, pretty cool.

![tmp2b8wnrq_](https://github.com/soeaver/M2FP/assets/23439776/1b0464b4-7800-4e78-95c7-5180f88f0a84)
